dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/AccountController.java

```diff
@@ -511,8 +515,27 @@ public ResponseEntity<Map<String, String>> updatePassword(
   }
 
   @GetMapping("/linkedAccounts")
-  public @ResponseBody List<User> getLinkedAccounts(@CurrentUser User currentUser) {
-    return userService.getLinkedUserAccounts(currentUser);
+  public @ResponseBody UserLookups getLinkedAccounts(@CurrentUser User currentUser) {
+    List<User> linkedUserAccounts = userService.getLinkedUserAccounts(currentUser);
+    List<UserLookup> userLookups =
+        linkedUserAccounts.stream().map(UserLookup::fromUser).collect(Collectors.toList());
+
+    for (int i = 0; i < linkedUserAccounts.size(); i++) {
+      userLookups
+          .get(i)
+          .setRoles(
+              linkedUserAccounts.get(i).getUserRoles().stream()
+                  .map(UserRole::getUid)
+                  .collect(Collectors.toSet()));
+      userLookups
+          .get(i)
+          .setGroups(
+              linkedUserAccounts.get(i).getGroups().stream()
+                  .map(UserGroup::getUid)
+                  .collect(Collectors.toSet()));
+    }
```

Comment:
Would be great if this can be pushed to service layer so we can have unit test or even better to the store layer with custom query to improve performance. Because `getUserRoles()` and `getGroups()` still trigger lazy select queries to database the performance is going to be affected if we have large collection. Another option is that the method `HibernateUserStore..getLinkedUserAccounts(currentUser)` could have `FetchType.EAGER`.

If we want to keep current solution then it seems there are some unnecessary for loops. Maybe we can rewrite as below:
```

UserLookups userLookups = new UserLookups();
linkedUserAccounts.foreach( u -> userLookups.add(UserLookUp.fromUser(user)));
```

Then  in `UserLookup.fromUser(user)`  we can set `roles` and `groups`
```
lookup.setDisplayName(displayName);
lookup.setRoles(user.getUserRoles().map(UserRole::getUid));
lookup.setGroups(user.getGroups().map(UserGroup::getUid));
```

Reply:
Good suggestion :+1:

Reply:
This is just a very quick way to make this work for linkedAccounts, which a special feature only in use by PEPFAR right now. I also doubt there will be many users returned as linked accounts.
The problem with setting roles and groups in the fromUser method is that we don't want to show roles and groups in the regular use case for UserInfo.

Reply:
This is just a very quick way to make this work for linkedAccounts, which a special feature only in use by PEPFAR right now. I also doubt there will be many users returned as linked accounts.
The problem with setting roles and groups in the fromUser method is that we don't want to show roles and groups in the regular use case for UserInfo.

Reply:
I guess we can make a special LinkedUser DTO object instead, to maybe make this cleaner.

Reply:
@netroms ok, then I think simply call `setRoles()` and `setGroups()` inside `UserLoopkup.fromUser(user)` instead of using the for loop would make this cleaner.

Reply:
That would not work in the other use cases when we don't want to have the roles and groups included.

Reply:
Ideally, we might just make a special SQL query to only return username, name, and role IDs and group IDs. Could be nice in other cases too.